server/query.go

```diff
@@ -0,0 +1,125 @@
+package server
+
+import (
+	"context"
+	"encoding/hex"
+	"errors"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/client"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	tm "github.com/tendermint/tendermint/proto/tendermint/types"
+	coretypes "github.com/tendermint/tendermint/rpc/core/types"
+)
+
+// QueryBlocksByEvents performs a search for blocks for a given set of events
+// via the Tendermint RPC. An event takes the form of:
+// "{eventAttribute}.{attributeKey} = '{attributeValue}'". Each event is
+// concatenated with an 'AND' operand. It returns a slice of Info object
+// containing blocks and metadata. An error is returned if the query fails.
+// If an empty string is provided it will order blocks by asc
+func QueryBlocksByEvents(clientCtx client.Context, events []string, page, limit int, orderBy string) (*sdk.SearchBlocksResult, error) {
+	if len(events) == 0 {
+		return nil, errors.New("must declare at least one event to search")
+	}
+
+	if page <= 0 {
+		return nil, errors.New("page must be greater than 0")
+	}
+
+	if limit <= 0 {
+		return nil, errors.New("limit must be greater than 0")
+	}
+
+	// XXX: implement ANY
+	query := strings.Join(events, " AND ")
+
+	node, err := clientCtx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: this may not always need to be proven
+	// https://github.com/cosmos/cosmos-sdk/issues/6807
+	resBlocks, err := node.BlockSearch(context.Background(), query, &page, &limit, orderBy)
+	if err != nil {
+		return nil, err
+	}
+
+	blocks, err := formatBlockResults(resBlocks.Blocks)
+	if err != nil {
+		return nil, err
+	}
+
+	result := sdk.NewSearchBlocksResult(int64(resBlocks.TotalCount), int64(len(blocks)), int64(page), int64(limit), blocks)
+
+	return result, nil
+}
+
+// get block by height
+func GetBlockByHeight(clientCtx client.Context, height *int64) (*tm.Block, error) {
+	// get the node
+	node, err := clientCtx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	// header -> BlockchainInfo
+	// header, tx -> Block
+	// results -> BlockResults
+	resBlock, err := node.Block(context.Background(), height)
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := mkBlockResult(resBlock)
+	if err != nil {
+		return out, err
+	}
+
+	return out, nil
+}
+
+func GetBlockByHash(clientCtx client.Context, hashHexString string) (*tm.Block, error) {
+	hash, err := hex.DecodeString(hashHexString)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the node
+	node, err := clientCtx.GetNode()
+	if err != nil {
+		return nil, err
+	}
+
+	resBlock, err := node.BlockByHash(context.Background(), hash)
+	if err != nil {
+		return nil, err
+	}
+
+	out, err := mkBlockResult(resBlock)
+	if err != nil {
+		return out, err
+	}
+
+	return out, nil
+}
+
+func mkBlockResult(resBlock *coretypes.ResultBlock) (*tm.Block, error) {
+	return sdk.NewResponseResultBlock(resBlock, resBlock.Block.Time.Format(time.RFC3339)), nil
```

Comment:
This panics 

```shell
simd query block --type=hash CCE0A913D8F5C861974D2CD4BA6DAEE4C7E3965B33C6446D00E14DCBBF39BB55
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x1c179bd]

goroutine 1 [running]:
github.com/cosmos/cosmos-sdk/server.mkBlockResult(0xc0015ac800)
        github.com/cosmos/cosmos-sdk@v0.47.0-rc1/server/query.go:110 +0x1d
github.com/cosmos/cosmos-sdk/server.GetBlockByHash({{0x0, 0x0, 0x0}, {0x30219b0, 0xc0015a70b0}, 0x0, {0xc00159a778, 0x4}, {0x3031778, 0xc0008a5840}, ...}, ...)
        github.com/cosmos/cosmos-sdk@v0.47.0-rc1/server/query.go:101 +0x14a
github.com/cosmos/cosmos-sdk/server.QueryBlockCmd.func1(0xc001493800?, {0xc00030a6e0, 0x1, 0x2?})
        github.com/cosmos/cosmos-sdk@v0.47.0-rc1/server/tm_cmds.go:268 +0x158
github.com/spf13/cobra.(*Command).execute(0xc001493800, {0xc00030a6a0, 0x2, 0x2})
        github.com/spf13/cobra@v1.6.1/command.go:916 +0x862
github.com/spf13/cobra.(*Command).ExecuteC(0xc001424300)
        github.com/spf13/cobra@v1.6.1/command.go:1044 +0x3bd
github.com/spf13/cobra.(*Command).Execute(...)
        github.com/spf13/cobra@v1.6.1/command.go:968
github.com/spf13/cobra.(*Command).ExecuteContext(...)
        github.com/spf13/cobra@v1.6.1/command.go:961
github.com/cosmos/cosmos-sdk/server/cmd.Execute(0x2057110?, {0x0, 0x0}, {0xc00013f398, 0x14})
        github.com/cosmos/cosmos-sdk@v0.47.0-rc1/server/cmd/execute.go:32 +0x179
main.main()
        cosmossdk.io/simapp/simd/main.go:14 +0x30
```

Reply:
Yeah, looks like the tendermint rpc returns a nil block with no error for a hash that doesn't exist.. I'm assuming that's not something we can fix so I'll patch it on our end.